### PR TITLE
feat(users): Email JWT blacklist

### PR DESCRIPTION
**Pull request by racnan**
## Type of Change

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
- Adding blacklisting for email tokens, this will prevent use of same email token to be used twice.
- Change Password & Reset Password logs out user from all open dashboard sessions.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Prevention of use of same email link to be used twice.
Forcing Users to re-enter password if password is changed.

## How did you test it?
With Dashboard Frontend.
Following links should not work twice:
- Magic Link
To generate magic link email use dashboard frontend or below curl.
```sh
curl --location --request POST '<URL>/user/connect_account' \
--header 'Content-Type: application/json' \
--data-raw '{
    "email": ""
}'
```
- Reset Password
To generate reset password email use dashboard frontend or below curl.
```sh
curl --location --request POST '<URL>/user/forgot_password' \
--header 'Content-Type: application/json' \
--data-raw '{
    "email": ""
}'
```

All sessions should be logged out after change password or reset password.

Expected error response for the use of email token twice
```json
{
    "error": {
        "type": "invalid_request",
        "message": "Invalid or expired link",
        "code": "UR_04"
    }
}
```

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [X] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/user.rs [hunk 1]
```diff
@@ -1050,12 +1069,15 @@ pub async fn verify_email_without_invite_checks(
     state: AppState,
     req: user_api::VerifyEmailRequest,
 ) -> UserResponse<user_api::DashboardEntryResponse> {
-    let token = auth::decode_jwt::<email_types::EmailToken>(&req.token.clone().expose(), &state)
-        .await
-        .change_context(UserErrors::LinkInvalid)?;
+    let token = req.token.clone().expose();
+    let email_token =
+        auth::decode_jwt::<email_types::EmailToken>(&req.token.clone().expose(), &state)
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:1074]:
```suggestion
        auth::decode_jwt::<email_types::EmailToken>(token, &state)
```

File: crates/router/src/core/user.rs [hunk 2]
```diff
@@ -1077,13 +1102,17 @@ pub async fn verify_email(
     state: AppState,
     req: user_api::VerifyEmailRequest,
 ) -> UserResponse<user_api::SignInResponse> {
-    let token = auth::decode_jwt::<email_types::EmailToken>(&req.token.clone().expose(), &state)
-        .await
-        .change_context(UserErrors::LinkInvalid)?;
+    let token = req.token.clone().expose();
+    let email_token =
+        auth::decode_jwt::<email_types::EmailToken>(&req.token.clone().expose(), &state)
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:1107]:
```suggestion
        auth::decode_jwt::<email_types::EmailToken>(token, &state)
```

**Review by ThisIsMani — COMMENTED** [review]
Minor suggestions

**Review by ThisIsMani — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]

File: crates/router/src/services/authentication/blacklist.rs [hunk 3]
```diff
@@ -47,6 +49,33 @@ pub async fn check_user_in_blacklist<A: AppStateInfo>(
...
+        .await
+        .change_context(UserErrors::InternalServerError)
+}
+
+#[cfg(feature = "email")]
+pub async fn check_email_token_in_blacklist(state: &AppState, token: &str) -> UserResult<()> {
+    let redis_conn = get_redis_connection(state).change_context(UserErrors::InternalServerError)?;
+    let blacklist_key = format!("{}{token}", EMAIL_TOKEN_BLACKLIST_PREFIX);
+    let key_exists = redis_conn
+        .exists::<()>(blacklist_key.as_str())
```
Comment (NishantJoshi00) [inline crates/router/src/services/authentication/blacklist.rs:69]: remove this

**Review by NishantJoshi00 — APPROVED** [review]

**Pull request merged**